pkg/controllers/cloud/networkpolicy.go

```diff
@@ -864,24 +878,91 @@ func (a *appliedToSecurityGroup) updateRules(r *NetworkPolicyReconciler) error {
 		a.hasRules = false
 		return nil
 	}
+	irules, erules := a.computeRules(nps)
+	if irules == nil || erules == nil {
+		return nil
+	}
+	r.Log.V(1).Info("AppliedToSecurityGroup update rules", "Name", a.id,
+		"ingressRules", irules, "egressRules", erules)
+	ch := securitygroup.CloudSecurityGroup.UpdateSecurityGroupRules(&a.id, irules, erules)
+	go func() {
+		err := <-ch
+		r.cloudResponse <- &securityGroupStatus{sg: a, op: securityGroupOperationUpdateRules, err: err}
+	}()
+	return nil
+}
+
+// computeRules compute all the ingress and egress rules of an appliedTo group.
+func (a *appliedToSecurityGroup) computeRules(nps []interface{}) ([]*securitygroup.IngressRule, []*securitygroup.EgressRule) {
 	irules := make([]*securitygroup.IngressRule, 0)
 	erules := make([]*securitygroup.EgressRule, 0)
 	for _, i := range nps {
 		np := i.(*networkPolicy)
 		if !np.rulesReady {
-			return nil
+			return nil, nil
 		}
 		irules = append(irules, deepcopy.Copy(np.ingressRules).([]*securitygroup.IngressRule)...)
 		erules = append(erules, deepcopy.Copy(np.egressRules).([]*securitygroup.EgressRule)...)
 	}
-	irules, erules = deduplicateIngressRules(irules), deduplicateEgressRules(erules)
-	r.Log.V(1).Info("AppliedToSecurityGroup update rules", "Name", a.id,
-		"ingressRules", irules, "egressRules", erules)
-	ch := securitygroup.CloudSecurityGroup.UpdateSecurityGroupRules(&a.id, irules, erules)
-	go func() {
-		err := <-ch
-		r.cloudResponse <- &securityGroupStatus{sg: a, op: securityGroupOperationUpdateRules, err: err}
-	}()
+	return deduplicateIngressRules(irules), deduplicateEgressRules(erules)
+}
+
+// updateAddrGroupReference updates appliedTo group references and notifies removed addrGroups that rules referencing them is removed.
+func (a *appliedToSecurityGroup) updateAddrGroupReference(r *NetworkPolicyReconciler) error {
```

Comment:
Can you verify what happens when addressGroup and AppliedToGroup are updated together. Would it still preserve your references. You can try it out with just updating appliedToGroup and verify if appliedTo update keeps the references intact

Reply:
Checked it works. It doesn't preserve the reference, but it will recompute the reference when new AT is created and update rule called.